Cleanup recorded glb export

TobiasCutik opened the pull request:
_Changes are used for ProjectCoordination Project_
- add new functions useful to finish recording
- fix issue where GLTFSceneImporter could not be successfully created from stream

I am not 100% sure about my GLTFSceneImporter.cs change... -> see comment

TobiasCutik committed:
add individual functions to end recording + create exporter

TobiasCutik committed:
skip dataLoader verification when creating a GLTFSceneImporter from stream

TobiasCutik committed:
Merge remote-tracking branch 'origin/customizable-animation-tracks' into cleanup-recorded-glb-export

TobiasCutik commented on Runtime/Scripts/GLTFSceneImporter.cs:
```diff
@@ -325,7 +325,8 @@ public GLTFSceneImporter(GLTFRoot rootNode, Stream gltfStream, ImportOptions opt
 
-			VerifyDataLoader();
+			_options = options;
+			//VerifyDataLoader();
```
On my branch I did just comment out the `VerifyDataLoader();`. On the latest `customizable-animation-tracks`, there was just no options assignment anymore and during merge I completely accepted the state of my `cleanup-recorded-glb-export` branch. I don't see why there should be no option assignment.

wschnepp replied:
I dont know these changes, so they are coming from the main repository. Maybe check their git history to figure out why this was changed in the first place.

Commenting-out methods seems like bad practice to me.

Either delete it altogether if you cannot see any reason to ever call the method, or maybe add a default method parameter that is set such that the `old` behavior is the default, and you explicitly override the default when calling

wschnepp commented on Runtime/Scripts/Timeline/GLTFRecorder.cs:
```diff
@@ -175,6 +175,47 @@ internal void endRecording(out Dictionary<Transform, AnimationData>? param)
+			// ensure correct animation pointer plugin settings are used
+			if (!recordAnimationPointer)
+				settings.ExportPlugins.RemoveAll(x => x is AnimationPointerExport);
+			else
+			if (!settings.ExportPlugins.Any(x => x is AnimationPointerExport))
```
either weird formatting or weird condition check

TobiasCutik replied:
I took this line/ code section 1:1 from ``public void EndRecording(Stream stream, string sceneName = "scene", GLTFSettings? settings = null)``, which is from the main repository. As my ``CreateAnimationExporterAfterRecording()`` has (at least partly) the same functionality, this section is functionality-wise correct. I don't know how to make it less weird. Please make a suggestion or we just leave it like this I guess?

wschnepp replied:
```suggestion
			else if (!settings.ExportPlugins.Any(x => x is AnimationPointerExport))
```

wschnepp commented on Runtime/Scripts/Timeline/GLTFRecorder.cs:
```diff
@@ -175,6 +175,47 @@ internal void endRecording(out Dictionary<Transform, AnimationData>? param)
+				+ "Total Keyframes: " + recordingAnimatedTransforms.Sum(x => x.Value.tracks.Sum(y => y.Values.Count())));
+			#endif
+
+			return true;
```
this method is weird to me.
How does this end the recording?
I assume by setting the `isRecording` property to false, which breaks some asynchronous loop?`

I think then i would choose a different name for the method, because this name suggests that it immediately ends recording and has error-like failure conditions, which does not seem to be the case.

I understand that the other already existing methods in the plugin are named similarly, but the code quality of the plugin is questionable at best, so i would prefer to use our code standards instead of the plugins` where possible.

What about `ScheduleEndRecording()`, or alternatively,
name this `StopRecording`, and rename the other methods to better reflect what they additionally do.

wschnepp replied:
Additionally, can you call this implementation in the other methods to reduce duplicate code?

TobiasCutik committed:
Update GLTFSceneImporter.cs

remove option assignment in public GLTFSceneImporter constructor (as it is already done in the private one)

TobiasCutik committed:
add optional bool verifyDataLoader to GLTFSceneImporter()

TobiasCutik committed:
rename some functions that (partially) end recording

wschnepp commented on Runtime/Scripts/Timeline/GLTFRecorder.cs:
```diff
@@ -175,7 +175,48 @@ internal void endRecording(out Dictionary<Transform, AnimationData>? param)
+			return new GLTFSceneExporter(new Transform[] { root }, exportContext);
+		}
+
+		public void EndRecordingAndSaveToStream(string filename, string sceneName = "scene", GLTFSettings? settings = null)
```
Which stream are we saving to? 👀 

This method has a file-path passed in, so it should be named accordingly

wschnepp commented on Runtime/Scripts/Timeline/GLTFRecorder.cs:
```diff
@@ -175,7 +175,48 @@ internal void endRecording(out Dictionary<Transform, AnimationData>? param)
+			return true;
+		}
+		
+		public GLTFSceneExporter CreateAnimationExporterAfterRecording(GLTFSettings? settings = null) 
```
Since the returned type is called "SceneExporter", i would prefer if the method name used the same naming, i.e. `CreateSceneExporterAfterRecording`.

wschnepp commented on Runtime/Scripts/Timeline/GLTFRecorder.cs:
```diff
@@ -175,7 +175,48 @@ internal void endRecording(out Dictionary<Transform, AnimationData>? param)
 		}
 
-		public void EndRecording(string filename, string sceneName = "scene", GLTFSettings? settings = null)
+		public bool TryEndRecordingDirect()
```
I feel like my original comment about this method still holds true?
If you disagree about my opinion, lets talk about it

TobiasCutik replied:
I kinda disagree/ dont see a better naming here. Lets talk

TobiasCutik committed:
simplify code + make renamings for clarity

TobiasCutik committed:
rename TryEndRecordingDirect() to EndRecording()
